service/grails-app/services/org/olf/rs/PatronNoticeService.groovy

```diff
@@ -0,0 +1,134 @@
+package org.olf.rs
+
+import com.k_int.okapi.OkapiClient
+import grails.core.GrailsApplication
+import grails.gorm.multitenancy.Tenants
+import groovy.json.JsonSlurper
+import org.apache.kafka.clients.consumer.KafkaConsumer
+
+import org.olf.rs.EmailService
+import org.olf.rs.EventConsumerService
+import org.olf.rs.EventPublicationService
+import org.olf.rs.NoticePolicyNotice
+
+public class PatronNoticeService {
+
+  EmailService emailService
+  EventConsumerService eventConsumerService
+  EventPublicationService eventPublicationService
+  OkapiClient okapiClient
+  GrailsApplication grailsApplication
+  private KafkaConsumer consumer = null
+
+  @javax.annotation.PostConstruct
+  public void init() {
+    // We need another consumer because EventConsumerService consumes all topics
+    // as they come in whereas the notice triggers need to queue until they can
+    // be consumed under the aegis of a timer request
+    log.debug("Configuring event consumer for patron notice service")
+    Properties props = new Properties()
+    try {
+      grailsApplication.config.events.consumer.toProperties().each { final String key, final String value ->
+        // Directly access each entry to cause lookup from env
+        String prop = grailsApplication.config.getProperty("events.consumer.${key}")
+        log.debug("Configuring event consumer for patron notice service :: key:${key} value:${value} prop:${prop}");
+        props.setProperty(key, prop)
+      }
+      log.debug("Configure consumer ${props}")
+    }
+    catch ( Exception e ) {
+      log.error("Problem assembling props for consume",e);
+    }
+
+    consumer = new KafkaConsumer(props)
+    log.debug("PatronNoticeService::init() returning");
+  }
+
+  @javax.annotation.PreDestroy
+  private void cleanUp() throws Exception {
+    log.info("PatronNoticeService::cleanUp");
+    consumer.wakeup();
+  }
+
+  public void triggerNotices(PatronRequest pr, String trigger) {
+    if (!(pr?.patronEmail instanceof String)) {
+      log.error("Notice triggered for request ${pr.hrid} with no email for patron");
+      return;
+    }
+    log.debug("triggerNotices(${pr.patronEmail}, ${trigger})")
+    String tenant = Tenants.currentId()
+    eventPublicationService.publishAsJSON(
+      "${tenant}_PatronNoticeEvents",
+      null,
+      [
+        event:'TriggerNotices_ind',
+        tenant: tenant,
+        payload:[
+          trigger: trigger,
+          email: pr.patronEmail,
+          values: [
+            user: [
+              id: pr.patronReference,
+              givenName: pr?.patronGivenName ?: "",
+              surame: pr.patronSurname,
+            ],
+            request: [
+              id: pr.hrid,
+              pickupLocation: "",
+              neededBy: ""
+              // pickupLocation: pr?.pickupLocation ?: "",
+              // neededBy: pr?.neededBy ?: ""
```

Comment:
Ok, I really don't know what to do here. For some reason my use of the Elvis operator upsets the serialisation of this map into JSON. But I checked in the GroovyConsole, it seems to work the way I think it does, it works above for givenName as it's not falling through, and a literal empty string works. But if I use the commented approach for those to fall back to empty string, THIS happens:
```
2020-11-19 16:17:28.072 ERROR --- [pool-1-thread-2] o.grails.events.EventSubscriberTrigger   : Error triggering event [PREventIndication] for subscriber [grails.events.subscriber.MethodSubscriber(public java.lang.Object org.olf.rs.ReshareApplicationEventHandle
rService.handleApplicationEvent(java.util.Map))]: null                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                    
java.lang.StackOverflowError: null                                                                                                                                                                                                                                  
        at java.base/java.util.concurrent.ConcurrentHashMap.putVal(ConcurrentHashMap.java:1012)                                                                                                                                                                     
        at java.base/java.util.concurrent.ConcurrentHashMap.putIfAbsent(ConcurrentHashMap.java:1541)                                                                                                                                                                
        at java.base/java.lang.ClassLoader.getClassLoadingLock(ClassLoader.java:668)                                                                                                                                                                                
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:591)                                                                                                                                                            
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:640)                                                                                                                                                            
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:616)                                                                                                                                                            
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:579)                                                                                                                                                                  
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)                                                                                                                                                               
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)                                                                                                                                                                                          
        at java.base/java.lang.Class.forName0(Native Method)                                                                                                                                                                                                        
        at java.base/java.lang.Class.forName(Class.java:315)                                                                                                                                                                                                        
        at groovy.lang.MetaClassRegistry$MetaClassCreationHandle.createWithCustomLookup(MetaClassRegistry.java:144)                                                                                                                                                 
        at groovy.lang.MetaClassRegistry$MetaClassCreationHandle.create(MetaClassRegistry.java:139)                                                                                                                                                                 
        at org.codehaus.groovy.reflection.ClassInfo.getMetaClassUnderLock(ClassInfo.java:286)                                                                                                                                                                       
        at org.codehaus.groovy.reflection.ClassInfo.getMetaClass(ClassInfo.java:329)                                                                                                                                                                                
        at org.codehaus.groovy.reflection.ClassInfo.getMetaClass(ClassInfo.java:339)                                                                                                                                                                                
        at org.codehaus.groovy.runtime.metaclass.MetaClassRegistryImpl.getMetaClass(MetaClassRegistryImpl.java:275)                                                                                                                                                 
        at org.codehaus.groovy.runtime.InvokerHelper.getMetaClass(InvokerHelper.java:965)                                                                                                                                                                           
        at org.codehaus.groovy.runtime.DefaultGroovyMethods.getMetaPropertyValues(DefaultGroovyMethods.java:533)                                                                                                                                                    
        at org.codehaus.groovy.runtime.DefaultGroovyMethods.getProperties(DefaultGroovyMethods.java:552)                                                                                                                                                            
        at groovy.json.DefaultJsonGenerator.getObjectProperties(DefaultJsonGenerator.java:242)                                                                                                                                                                      
        at groovy.json.DefaultJsonGenerator.writeObject(DefaultJsonGenerator.java:236)                                                                                                                                                                              
        at groovy.json.DefaultJsonGenerator.writeObject(DefaultJsonGenerator.java:164)                                                                                                                                                                              
        at groovy.json.DefaultJsonGenerator.writeIterator(DefaultJsonGenerator.java:402)                                                                                                                                                                            
        at groovy.json.DefaultJsonGenerator.writeObject(DefaultJsonGenerator.java:202)                                                                                                                                                                              
        at groovy.json.DefaultJsonGenerator.writeMapEntry(DefaultJsonGenerator.java:387)                                                                                                                                                                            
        at groovy.json.DefaultJsonGenerator.writeMap(DefaultJsonGenerator.java:375)                                                                                                                                                                                 
        at groovy.json.DefaultJsonGenerator.writeObject(DefaultJsonGenerator.java:237)                                                                                                                                                                              
        at groovy.json.DefaultJsonGenerator.writeObject(DefaultJsonGenerator.java:164)                                                                                                                                                                              
        at groovy.json.DefaultJsonGenerator.writeIterator(DefaultJsonGenerator.java:402)                                                                                                                                                                            
        at groovy.json.DefaultJsonGenerator.writeObject(DefaultJsonGenerator.java:202)                                                                                                                                                                              
        at groovy.json.DefaultJsonGenerator.writeMapEntry(DefaultJsonGenerator.java:387)                                                                                                                                                                            
        at groovy.json.DefaultJsonGenerator.writeMap(DefaultJsonGenerator.java:375)                                                                                                                                                                                 
        at groovy.json.DefaultJsonGenerator.writeObject(DefaultJsonGenerator.java:237)                                                                                                                                                                              
        at groovy.json.DefaultJsonGenerator.writeObject(DefaultJsonGenerator.java:164)                                                                                                                                                                              
        at groovy.json.DefaultJsonGenerator.writeIterator(DefaultJsonGenerator.java:402)                                                                                                                                                                            
        at groovy.json.DefaultJsonGenerator.writeObject(DefaultJsonGenerator.java:202)                                                                                                                                                                              
        at groovy.json.DefaultJsonGenerator.writeMapEntry(DefaultJsonGenerator.java:387)
        ....this repeats for a few screens
        at groovy.json.DefaultJsonGenerator.writeIterator(DefaultJsonGenerator.java:402)                                                                                                                                                                            
        at groovy.json.DefaultJsonGenerator.writeObject(DefaultJsonGenerator.java:202)                                                                                                                                                                              
        at groovy.json.DefaultJsonGenerator.writeMapEntry(DefaultJsonGenerator.java:387)                                                                                                                                                                            
        at groovy.json.DefaultJsonGenerator.writeMap(DefaultJsonGenerator.java:375)                                                                                                                                                                                 
        at groovy.json.DefaultJsonGenerator.writeObject(DefaultJsonGenerator.java:237)         
```

Reply:
One thing I'd watch out for is using "" instead of '' = "" will give a GString wheras '' gives a String. Some JSON serializers don't do sensible things with GString so either

    a: null?.why?:''
or
    a: null?.why?:"".toString()

would be better